boards/arm/cc1352p1_launchxl/board_antenna.c

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2021 Florin Stancu
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * Implements the RF driver callback to configure the on-board antenna
+ * switch.
+ */
+
+#define DT_DRV_COMPAT ti_antenna_mux
+
+#include <zephyr/init.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/pinctrl.h>
+
+#include <ti/drivers/rf/RF.h>
+#include <driverlib/rom.h>
+#include <driverlib/interrupt.h>
+
+/* custom pinctrl states for the antenna mux */
+#define PINCTRL_STATE_ANT_24G		1
+#define PINCTRL_STATE_ANT_24G_PA	2
+#define PINCTRL_STATE_ANT_SUBG		3
+#define PINCTRL_STATE_ANT_SUBG_PA	4
+
+#define BOARD_ANT_GPIO_24G   0
+#define BOARD_ANT_GPIO_PA    1
+#define BOARD_ANT_GPIO_SUBG  2
+
+#define ANTENNA_MUX DT_NODELABEL(antenna_mux0)
+
+static int board_antenna_init(const struct device *dev);
+static void board_cc13xx_rf_callback(RF_Handle client, RF_GlobalEvent events,
+		void *arg);
+
+const RFCC26XX_HWAttrsV2 RFCC26XX_hwAttrs = {
+	.hwiPriority        = INT_PRI_LEVEL7,
+	.swiPriority        = 0,
+	.xoscHfAlwaysNeeded = true,
+	/* RF driver callback for custom antenna switching */
+	.globalCallback = board_cc13xx_rf_callback,
+	/* Subscribe to events */
+	.globalEventMask = (RF_GlobalEventRadioSetup |
+			RF_GlobalEventRadioPowerDown),
+};
+
+PINCTRL_DT_INST_DEFINE(0);
+DEVICE_DT_INST_DEFINE(0, board_antenna_init, NULL, NULL, NULL,
+					  POST_KERNEL, CONFIG_BOARD_ANTENNA_INIT_PRIO, NULL);
+
+static const struct pinctrl_dev_config *ant_pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(0);
+static const struct gpio_dt_spec ant_gpios[] = {
+	GPIO_DT_SPEC_GET_BY_IDX_OR(ANTENNA_MUX, gpios, BOARD_ANT_GPIO_24G, {0}),
+	GPIO_DT_SPEC_GET_BY_IDX_OR(ANTENNA_MUX, gpios, BOARD_ANT_GPIO_PA, {0}),
+	GPIO_DT_SPEC_GET_BY_IDX_OR(ANTENNA_MUX, gpios, BOARD_ANT_GPIO_SUBG, {0}),
+};
+
+
+/**
+ * Antenna switch GPIO init routine.
+ */
+int board_antenna_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	int i;
+
+	/* default pinctrl configuration: set all antenna mux control pins as GPIOs */
+	pinctrl_apply_state(ant_pcfg, PINCTRL_STATE_DEFAULT);
```

Comment:
this all continues to look like a hack to me.

Reply:
The antenna switch pins must either be asserted as GPIOs or re-mapped to the chip's radio core based on the current RF configuration scenario. Thus, Zephyr's GPIO & PinCtrl APIs must be simultaneously used to re-arrange the antenna pathways. Although it might seem a bit hackish since it is not the intended use case of this API (though it seems okay for me), it was done to avoid using the low-level HAL routines (as requested earlier in this PR).